2단계 - @MVC 구현

kang-hyungu opened the pull request:
구현이 금방 끝날 줄 알았는데 생각보다 오래 걸렸네요 🙄 
코드 리뷰 부탁드립니다!

kang-hyungu committed:
feat : 요구사항 1 - 애노테이션 기반 MVC 프레임워크

- @Controller 애노테이션이 설정되어 있는 클래스를 찾은 후 @RequestMapping 설정에 따라 요청 URL과 메소드를 연결하도록 구현한다.

kang-hyungu committed:
feat : 요구사항 2 - 레거시 MVC와 애노테이션 기반 MVC 통합

- controller 인터페이스 상속과 controller 어노테이션 기반 mvc 통합 가능하도록 수정

javajigi commented on src/main/java/core/annotation/web/RequestMapping.java:
```diff
@@ -10,5 +10,5 @@
     String value() default "";
 
-    RequestMethod method() default RequestMethod.GET;
+    RequestMethod[] method() default {};
```
오호. 배열까지 확장 👍

javajigi commented on src/main/java/core/mvc/asis/DispatcherServlet.java:
```diff
@@ -18,28 +24,62 @@
     private static final String DEFAULT_REDIRECT_PREFIX = "redirect:";
 
     private RequestMapping rm;
+    private AnnotationHandlerMapping annotationHandlerMapping;
```
RequestMapping과 AnnotationHandlerMapping이 요청에 대한 Controller를 반환하는 것으로 같은 역할을 하고 있다.
인터페이스를 추가하고, 이에 따른 2개의 구현체를 추가하는 구조로 개발해 본다.

javajigi commented on src/main/java/core/mvc/asis/DispatcherServlet.java:
```diff
@@ -10,6 +15,7 @@
+import java.util.Map;
 
 @WebServlet(name = "dispatcher", urlPatterns = "/", loadOnStartup = 1)
 public class DispatcherServlet extends HttpServlet {
```
DispatcherServlet은 서블릿 컨테이너가 생성하고 초기화를 담당한다.
따라서 DI를 통해 RequestMapping, AnnotationHandlerMapping에 대한 의존관계를 구현하기 힘들다.
이와 같이 DI를 통해 의존관계를 연결하기 힘든 구조의 경우 어떤 구조로 구현하면  RequestMapping, AnnotationHandlerMapping에 대한 의존 관계를 분리할 수 있을까?

kang-hyungu replied:
의존관계를 주입할 클라이언트를 구현하기 힘든 상황이라면 외부 설정파일에서 의존관계를 결정하거나 프록시로 감싸서 분리 해볼 수 있을 것 같습니다.

javajigi commented on src/main/java/core/mvc/asis/DispatcherServlet.java:
```diff
@@ -18,28 +24,62 @@
         rm = new RequestMapping();
         rm.initMapping();
+
+        annotationHandlerMapping = new AnnotationHandlerMapping(WebServerLauncher.class);
```
WebServerLauncher를 굳이 전달하기보다 scan하고 싶은 package를 문자열로 전달하는 방식으로 구현해 본다.

javajigi commented on src/main/java/core/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -1,27 +1,93 @@
+                        new SubTypesScanner(),
+                        new FieldAnnotationsScanner(),
+                        new MethodAnnotationsScanner(),
+                        new TypeAnnotationsScanner())
```
Controller scan경우 TypeAnnotationsScanner만 있으면 되지 않을까?

javajigi commented on src/main/java/next/controller/MyController.java:
```diff
@@ -0,0 +1,41 @@
+import javax.servlet.http.HttpServletResponse;
+
+@Controller
+public class MyController {
```
기존 MVC에서 애노테이션 기반(@MVC)와 병행해서 동작하도록 구현했다.
기존 Controller를 새롭게 구현한 @MVC와 병행해서 동작하는지 확인하기 위해 기존 Controller의 일부를 @MVC 기반 Controller로 수정한다.
이와 같이 이동해 봄으로써 asis MVC 프레임워크와 새롭게 구현한 tobe MVC 프레임워크가 동시에 동작하도록 구현해 본다.

javajigi reviewed (changes requested):
@MVC 프레임워크가 생각보다 구현하기 쉽지 않을텐데 구현 잘 했네요. 💯 
추가 연습과 리팩토링을 위한 피드백 몇 개 남겼으니, 3단계 리팩토링 힌트 참고해 같이 반영해 보세요.

kang-hyungu committed:
feat : 요구사항 2 - 피드백 반영

- Controller scan은 TypeAnnotationsScanner만 사용하도록 수정
- AnnotationHandlerMapping에 basePackage를 클래스가 아닌 문자열로 지정할 수 있도록 변경

kang-hyungu committed:
feat : 2단계 - 피드백 반영

- 레거시 MVC와 @MVC가 병행 동작 가능한지 확인 할 수 있도록 CreateUserController, UpdateUserController, ListUserController를 UserController에 구현

kang-hyungu committed:
feat : 2단계 - 피드백 반영

- RequestMapping을 LegacyHandlerMapping으로 변경
- HandlerMapping 인터페이스 추가하여 LegacyHandlerMapping, AnnotationHandlerMapping로 확장

javajigi commented on src/main/java/core/mvc/asis/Controller.java:
```diff
@@ -1,8 +1,10 @@
 
 public interface Controller {
-    String execute(HttpServletRequest req, HttpServletResponse resp) throws Exception;
+    ModelAndView execute(HttpServletRequest req, HttpServletResponse resp) throws Exception;
```
기존의 Controller를 그대로 유지하면서 공존 가능한 구조를 찾아야하지 않을까?
기존의 Controller 구조를 변경하는 것 또한 리팩토링에 위험요소이지 않을까? 또는 기존의 Controller 구조를 변경할 수 없는 상태라면 어떻게 될까?

kang-hyungu replied:
레거시 코드를 유지해야된다는 조건을 깜빡했네요;
다시 수정했습니다!

javajigi commented on src/main/java/core/mvc/asis/DispatcherServlet.java:
```diff
@@ -47,47 +41,55 @@ protected void service(HttpServletRequest req, HttpServletResponse resp) throws
-        if (viewName.startsWith(DEFAULT_REDIRECT_PREFIX)) {
-            resp.sendRedirect(viewName.substring(DEFAULT_REDIRECT_PREFIX.length()));
-            return;
+    private ModelAndView getModelAndView(HttpServletRequest req, HttpServletResponse resp, Object handler) throws Exception {
```
프레임워크를 구현하는 관점에서 봤을 때 Servlet, Struts1, Struts2와 같은 다양한 프레임워크와의 통합을 하고 싶다.
이와 같이 새로운 기술에 대한 통합을 시도할 경우 if/else가 계속해서 추가될 수 있다.
이와 같이 if/else가 늘어날 것으로 예상되는 부분에 자바의 다형성을 활용해 interface를 추가하고 확장 가능한 구조로 개선해 보면 어떨까?

```
public interface HandlerAdapter {
    boolean supports(Object handler);

    ModelAndView handle(
        HttpServletRequest request, 
        HttpServletResponse response, 
        Object handler) throws Exception;
}
```

kang-hyungu replied:
HandlerAdapter 인터페이스로 추상화해서 레거시와 어노테이션용으로 확장했습니다.
서블릿 init()에서 계속 의존관계를 주입하게 되니 한 곳에서 관리하는 장점이 있지만 이전에 피드백 남기신 것처럼 의존 관계 분리가 어려워지겠네요.

javajigi reviewed (approved):
피드백 반영 잘 했네요. 👍 
단, Controller의 반환 값은 String으로 그대로 유지한 상태에서 공존 가능하도록 구현해 보면 좋겠어요.
피드백 남겼으니 참고해 보세요. 3단계 힌트 내용까지 포함해서 리팩토링 진행해 보면 좋겠어요.
